test/integration/secrets_test_cases.py

```diff
@@ -0,0 +1,157 @@
+# Copyright 2025 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import ast
+import inspect
+import textwrap
+from typing import Any, Callable, NamedTuple
+from unittest.mock import ANY
+
+import jubilant
+
+import ops
+import ops.testing
+from test.charms.test_secrets.src.charm import TestSecretsCharm
+
+UnitCode = Callable[[TestSecretsCharm, dict[str, Any]], None]
+"""
+Python code that will be executed in the test charm is defined
+as a function with this type.
+
+Arguments names must be exactly these, if used:
+    self: TestSecretsCharm
+    rv: dict[str, Any]
+
+Example:
+
+def code(self: TestSecretsCharm, rv: dict[str, Any]):
+    '''Test get_secret by id.'''
+    self.model.get_secret(id=self._stored.secret_id)
+    rv['_result'] = 42
+"""
+
+ScenarioAssertions = Callable[[ops.testing.Secret | None, dict[str, Any] | None], None]
+"""
+Validate the result in Scenario.
+
+Example:
+
+assert result == {
+    '_before': None,
+    '_after': {
+        'info': {
+            'id': ANY,
+            'label': None,
+            'revision': 1,
+            'expires': None,
+            'rotation': None,
+            'rotates': None,
+            'description': None,
+        },
+        'tracked': {'foo': 'bar'},
+        'latest': {'foo': 'bar'},
+    },
+}
+
+assert secret
+assert secret.owner == 'application'
+assert not secret.remote_grants
+"""
+
+JubilantAssertions = Callable[[jubilant.RevealedSecret | None], None]
+"""
+Validate the result in Jubilant.
+"""
+
+
+# Note: this can't be called e.g. `TestCase` as pytest would trip
+class Case(NamedTuple):
+    func: UnitCode
+    scenario_assertions: ScenarioAssertions | None
+    jubilant_assertions: JubilantAssertions | None
+    # Later:
+    # - jubilant_assertions
+    # - setup?
+    # - cleanup?
+
+    @property
+    def code(self) -> str:
+        func_src = inspect.getsource(self.func)
+        func_def = ast.parse(func_src).body[0]
+        assert isinstance(func_def, ast.FunctionDef)
+        return '\n'.join(textwrap.dedent(ast.unparse(stmt)) for stmt in func_def.body)
+
+
+def add_secret(self: TestSecretsCharm, rv: dict[str, Any]):
+    """Add secret with content."""
+    secret: ops.Secret = self.app.add_secret({'foo': 'bar'})
```

Comment:
It seems to this entire complicated setup exists to reuse this one line of a code, a simple `add_secret` call. I know we may add test cases in future, but I think we should revert to a KISS approach, with completely separate tests in `test_e2e` (unit) and `test_secrets` (integration), with no shared code. If later on it gets to be a *lot* of duplication we can evaluate the best way to share.